com.microsoft.mrtk.core/Utilities/XRSubsystemHelpers.cs

```diff
@@ -51,53 +51,56 @@ public static class XRSubsystemHelpers
             runningSubsystems.RemoveAll(subsystem => !subsystem.running);
         }
 
         /// <summary>
-        /// Get all running subsystems of type T.
+        /// Get all running subsystems of type T. Note, this is internally cached, so
+        /// the list is only valid immediately after invocation.
         /// </summary>
         /// <remarks>
-        /// Caution: this method allocs a new list of subsystems.
-        /// For performance critical (frame loop) applications,
-        /// consider calling SubsystemManager.GetSubsystems directly
-        /// with a pre-allocated list.
+        /// This method allocates on the first invocation with a new, unique type T.
+        /// Subsequent invocations with the same type T will no longer allocate.
         /// </remarks>
         public static List<T> GetAllRunningSubsystems<T>() where T : ISubsystem
         {
-            List<T> filtered = new List<T>();
-
-            foreach (T sys in GetAllSubsystems<T>())
-            {
-                if (sys.running)
-                {
-                    filtered.Add(sys);
-                }
-            }
-
-            return filtered;
+            List<T> subsystems = GetAllSubsystems<T>();
+            subsystems.RemoveAll(subsystem => !subsystem.running);
+            return subsystems;
         }
 
         /// <summary>
-        /// Get all subsystems (not necessarily running) of type T.
+        /// Get all subsystems (not necessarily running) of type T. Note, this is
+        /// internally cached, so the list is only valid immediately after invocation.
         /// </summary>
         /// <remarks>
-        /// Caution: this method allocs a new list of subsystems.
-        /// For performance critical (frame loop) applications,
-        /// consider calling SubsystemManager.GetSubsystems directly
-        /// with a pre-allocated list.
+        /// This method allocates on the first invocation with a new, unique type T.
+        /// Subsequent invocations with the same type T will no longer allocate.
         /// </remarks>
         public static List<T> GetAllSubsystems<T>() where T : ISubsystem
         {
-            List<T> subsystems = new List<T>();
-            SubsystemManager.GetSubsystems(subsystems);
-            return subsystems;
+            // If we haven't queried subsystems of this type before,
+            // we'll need to add a new entry in our cache. This will
+            // allocate the first time, but subsequent queries will
+            // reuse the same cached list. (The list will still be cleared
+            // and re-filled by the GetSubsystems call, but nothing will be
+            // allocated.)
+            if (!scratchCaches.ContainsKey(typeof(T)))
+            {
+                scratchCaches.Add(typeof(T), new List<T>());
+            }
+
+            List<T> cache = scratchCaches[typeof(T)] as List<T>;
+            SubsystemManager.GetSubsystems<T>(cache);
```

Comment:
Personally, I would like to see this element performance timed, unless there is guidance from the Unity team that ```GetSubsystems<T>``` does not reinitialise the List or do a duplicate check against the systems, else this "scratchCaches" is just adding an overhead that is not required.

Reply:
The exact implementation can be found at https://github.com/Unity-Technologies/UnityCsReference/blob/a935c76939be457b82c70f54fe2cc7dd40fb9090/Modules/Subsystems/SubsystemManager.cs#L26-L44

It simply clears the list and conditionally copies subsystems out of the various internal caches. This change saves us a per-call allocation of the list we pass in.

Reply:
I would argue based on the code provided @keveleigh , this the above change does little to help / cache anything.

```csharp
        public static void GetSubsystems<T>(List<T> subsystems)
            where T : ISubsystem
        {
            subsystems.Clear();
            AddSubsystemSubset(s_IntegratedSubsystems, subsystems);
            AddSubsystemSubset(s_StandaloneSubsystems, subsystems);
            AddSubsystemSubset(s_DeprecatedSubsystems, subsystems);
        }
```
As stated, the LIST that is passed in is Cleared prior to the SubSet queries being run.

So NOTHING that you pass in is preserved, so the cache is actually retaining redundant information.  All it "possibly" is saving you is the "New" initialisation of a List which is so minor to wouldn't be worth having the cache above.

As stated, I'd like to see timings to prove the cache saves anything, which IMHO, given the frequency that you are likely to call "GetSubSystems" is very low, the benefit is very minor.

Realistically, if you want to use the cache effectively, then cache the return from the GetSubSystems call and DO NOT call GetSubSystems again during the life of the app or have cache invalidation to reduce the calls.
But the above functionality is only adding overhead for a very minor benefit.

Reply:
P.S. @keveleigh , thanks for reminding me the CSReference still exists and is even being kept up to date :D

Reply:
>All it "possibly" is saving you is the "New" initialisation of a List which is so minor to wouldn't be worth having the cache above.

This is exactly what it's saving, and I'd argue _not_ minor at all. Per-call (so, potentially multiple times per-frame) allocations can add up quickly.

> Realistically, if you want to use the cache effectively, then cache the return from the GetSubSystems call and DO NOT call GetSubSystems again during the life of the app or have cache invalidation to reduce the calls.
But the above functionality is only adding overhead for a very minor benefit.

Subsystem can be registered, started, and stopped at runtime, so I don't think caching the result of this call is a good idea.

I agree that profiling this would be good to see what changes on timing (since a dictionary lookup could add some time), but I guarantee it saves a per-call allocation of memory, so if timing is reasonably similar, there's only benefit here.

Reply:
OK, after a little more research and testing, caching the List might be a good idea, not to save the new() call however, but to preserve the native array that is inherent within the List Type, reducing the GC load and performance hit of adding more items than a List<T> holds as the ```GetSubsystems``` uses.  

However, this could also be achieved by caching the count of the systems and using the overloaded version of new <T> with the index size, which would reduce the GC hit from the List being cleared.

Ultimately though, if this is being called every frame and it is extremely unlikely that the contents will change between frames, then using the cache for the same frame would be more performant than calling GetSubsystems exclusively.  

> Subsystem can be registered, started, and stopped at runtime, so I don't think caching the result of this call is a good idea.

If this truly is the case, then it would be FAR more performant to only use GetSubsystems when such a change occurs, which should be easily trackable, either by simply examining the count of systems, or better, working with Unity to have events on the Subsystems for changes.

Reply:
Did some profiling. 1000 calls to the `HandsAggregator` helper property in `XRSubsystemHelpers` every frame registers as 0.00ms.

Reply:
Detailed profiling results below.